source/extensions/filters/http/jwt_authn/verifier.cc

```diff
@@ -257,7 +254,12 @@ VerifierPtr innerCreate(const JwtRequirement& requirement,
   if (it == providers.end()) {
     throw EnvoyException(fmt::format("Required provider ['{}'] is not configured.", provider_name));
   }
-  return std::make_unique<ProviderVerifierImpl>(audiences, factory, it->second, parent);
+  if (audiences.empty()) {
+    return std::make_unique<ProviderVerifierImpl>(nullptr, factory, it->second, parent);
```

Comment:
For normal Provider,  ( not provider_with_audience),  the audiences should come from the Provider.   
Here: https://github.com/envoyproxy/envoy/blob/master/api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto#L55

Provider_with_audience just a way for users to override it.   

1) So ProviderVerifierImpl can always have audiences, So we can pass const CheckAudienc& to Auth.

2) CheckAudiences object in JwksDataImpl is not needed any more. can be removed.

Reply:
On second thought, allow failed or missing can have more than providers. Should this be a map of issuers to check audiences?

Reply:
No,  it is not a 1 to 1 mapping between issuer and a set of audiences.   

In the shape of current config,  multiple providers may have same issuer but with different audiences.  or other different fields, such as from_headers, forward_payload_header.

provider1: 
    issuser: A
    audiences: A1
provider2: 
    issuser: A
    audiences: A2

The audiences in the JwksCache is just for "allow_fai"l case since it could not specify provider.    When JWT token arrives, find the audiences by the  issuer in the JwksCache, and use it.
For the above example, provider2 will be in the JwksCache.   So for allow_fail,  it is better not to have different providers with same issuer and with different audiences.  Anyway,  allow_fail is a very special use case for Istio authentication.

But if "rules" is provided, this is OK.   Rule can specify Provider, everything should work.  The key is audiences and other fields should be stored in ProviderVerifierImpl. 

Since Extractor is created in ProviderImpl so "from_header" and "from_param" are already per-provider. which is good.

The same goes with other two fields:  "forward" and "forward_payload_header" should be stored in ProviderImpl too. and pass them to Auth.  

Maybe define a ProviderData which includes CheckAudience object,  and all other fields. Store one in ProviderImpl and one in JwksCacheData.  What do you think?

Reply:
I think it's better to encapsulate issuer, audiences, forward and forward_payload_header in a separate class something like 
```
class ProviderChecker {
public:
  virtual ~ProviderChecker() {}
  
  // First verifier the issuer is valid, then check the audiences. Can return JwtUnknownIssuer and
  // JwtAudienceNotAllowed.
  virtual ::google::jwt_verify::Status
  checkIssuerAndAudience(const std::string& issuer,
                         const std::vector<std::string>& jwt_audiences) const PURE;
  
  // If JWT payload should be forwarded or not 
  virtual bool shouldForward() const PURE;

  // If JWT payload should be forwarded, the header to store the payload.
  virtual const std::string& forwardPayloadHeader() const PURE;
};
```
provider verifier will have an unique_ptr to one impl where it only have one issuer and check audience. Allow failed will have another impl where there is a map of potentially many issuers to check audiences. References to ProviderChecker is passed to auth.

Unrelated question, should forward rewrite the query param too?

Reply:
Currently, If there are 2 different providers with the same issuer, the only the audiences in the last defined provider is valid? Should we concat the audience lists for allow missing failed?

Reply:
Ideally,  JwksCache should be keyed by provider_name.  But for allow_fail case, we don't have a provider name,  we only have issuer after parsed JWT token.  So one way to support allow_fail is: to have JwksCache to be keyed by both provider_name and issuer.  By doing this way, we don't need to store these fields in ProviderImpl again,  just store them in JwksCache as the current code now.

We need to pass provide_name to Auth if it is from ProviderImpl so it can be used to lookup JwksCache.

The detail we have to figure out is:  if there are multiple providers with same issuer.  which one should the key issuer points to in the JwksCache.  It doesn't matter.  It is only used for allow_fail, it can point to any provider.

We can go with this approach.

Reply:
OK

Reply:
What about "forward" with from_query? Should we rewrite the query to remove JWT?

Reply:
No,  it is not easy to re-write query.   Let us not worry about it for now.  Add TODO.
For header, it is easy.  Let us remove it if forward = false.